README.md
# PSModule.Base

PowerShell - A personal module. Using an already existing one to (partly) migrate to and get familiar with Git.

To see a list of changes in this version, refer to the [Changelog](CHANGELOG.md).

## My first (real) attempt to

+ Setup and use GitHub.
+ Get familiar with and use [ModuleBuilder](https://github.com/PoshCode/ModuleBuilder) to prepare to create a module.
  + A way of working to enable building a (_module_).psm1 and use versioning (if required).
+ So this repo only contains the files to *build* the module. You need to get familiar with _ModuleBuilder_ and (re)create the module in your own environment.

Thoughts:

+ Keep an eye on the [PSModulePath issue](https://github.com/PowerShell/PowerShell/issues/6850) between PowerShell and PowerShell Core. So putting modules in a 'fixed' place and loading them might need tweaking in your personal environment. There are more PSModulePath things to consider. See the PowerShell GitHub.
+ Usage should be possible in both *PowerShell* and *PowerShell Core*.

## Prerequisites

The current _Build.ps1_ points to an output folder for the final module. In my case to a folder in my _$env:OneDrive_ location. Make sure that folder exists and preferably the main module folder is in your _$env:PSModulePath_.

## Some Markdown guides

+ [Markdown guide](https://www.markdownguide.org/basic-syntax/)
+ [John Gruder - markdown creator](https://daringfireball.net/projects/markdown/)
+ [GitHub mastering markdown](https://guides.github.com/features/mastering-markdown/)

## Build Module

Requires the already loaded module *ModuleBuilder* as described at the beginning. I'm not explaining the ModuleBuilder but only give a quick summary how to build the final module. People who use ModuleBuilder will know what to change. For others I'll advise to use all the defaults.

After you did a:

```PowerShell
git clone https://github.com/gitoldi/PSModule-Base.git
set-location PSModule-Base
```

The tree will be created under the current folder you started the command. You might want to modify 'sources\build.psd1'. When running the 'Build-Module' it creates the module in one of the standard PowerShell [$env:PSModulePath](https://docs.microsoft.com/en-us/powershell/developer/module/modifying-the-psmodulepath-installation-path) folders.

The default place will be: $( $env:USERPROFILE )\Documents\WindowsPowerShell\Modules\PSModule-Base

If you want the module in another PowerShell (Core) personal, group or system folder you can modify the 'OutputDirectory' variable. For group or system folder make sure you have the proper (admin) rights. The default setup will create the module in the (sub)folders and files in the given top folder 'PSModule-Base'.

```PowerShell
PS> Build-Module <full-path-to>\PSModule-Base\Sources\ -Prefix prefix.ps1 -Suffix suffix.ps1
```

Now the module is created and you'll find it in the folder as described in the 'build.psd1'.

```PowerShell
PS> Get-ChildItem $env:USERPROFILE'\documents\windowspowershell\modules\psmodule-personal'

Directory: C:\Users\<username>\documents\windowspowershell\modules\psmodule-personal

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        10/6/2019   5:01 PM                en-US
d-----        10/6/2019   5:01 PM                nl-NL
-a----        10/4/2019   8:27 PM           2479 PSModule-Base-Config-Template.psd1
-a----        10/6/2019   5:01 PM           2493 PSModule-Base.psd1
-a----        10/6/2019   5:01 PM          16747 PSModule-Base.psm1
```

+ File : PSModule-Base.psd1
  + The module manifest file.
+ File : PSModule-Base.psm1
  + The module file is created by the *Build-Module* and is the concatenation of:
    + Prefix.ps1
      + As supplied in the *Build-Module* command '-prefix prefix.ps1'.
        + e.g. This locates and reads the config file.
      + All functions
        + All functions defined in the *private* and *public* folders.
      + Suffix.ps1
        + As supplied in the *Build-Module* command '-suffix suffix.ps1'.
          + e.g. Anything you want to do at the end of importing the module.
+ File : PSModule-Base-Config-Template.psm1
  + x

## Import module

The module must be found in one of the standard module paths for *PowerShell* or *PowerShell Core*.

Since importing modules depends on your personal module tree and setup, in this README I use simple commands.
Personally I link PowerShell (folder WindowPowerShell and file Microsoft.PowerShell_profile.ps1) and PowerShell Core (folder PowerShell and file Profile.ps1) together.

```PowerShell
PS> Import-Module PSModule-Base
```

## Folders and files in this repo

In the root of the repo:

+ File - CHANGELOG.md
  + This file.
  + Keep all modifcations in this file? That is all ps1, psd1, psm1, ...???
+ File - README.md (this file)
  + Generic explanation of this module.
+ File - LICENSE
  + Current module license.
  + GNU GENERAL PUBLIC LICENSE Version 3.

### Build.ps1

### Prefix.ps1

### Suffix.ps1

### PSModule-Base.psd1

### PSModule-Base-Config-Template.psd1

### PSModule-Base.[On, Off]

## Folder - Source</h1>

### Subfolder - en-US

A small English about.
Here some more work is required, maybe explain the way to work with multiple languages.

### Subfolder - nl-NL

Een kleine Nederlandse uitleg.
Hier is meer werk noodzakelijk, misschien uitleggen hoe te werken met meerdere talen.

### Folder - Sources

This folder contains the main files to create the final module.

+ File - [Build.psd1](#build)
  + Configuration file for ['ModuleBuilder'](https://github.com/PoshCode/ModuleBuilder)
  + Trying to combine some suggested way's of working into my module.
+ File - [Prefix.ps1](#prefix)
  + Used by ModuleBuilder.
    + The contents of this file will be put at the beginning of the created (_module_).psm1 file containing all functions.
    + Here you can do some pre-requisite actions for your module. e.g.: Read a configuration file.
+ File - [Suffix.ps1](#suffix)
  + Used by ModuleBuilder.
    + The contents of this file will be put at the end of the created (_module_).psm1 file containing all functions.
    + Here you can do some cleanup or start something after the main part of your module is loaded.
+ File - PSModule-Base.psd1
  + The manifest file for the module.
+ File - PSModule-Base-Config.psd1
  + An example PowerShell configuration file for a module.

### Subfolder - Private

### Subfolder - Public

### Subfolder - Tests

### File - Build.psd1

### File - Tools.psd1

## Tested on

Operation System | PowerShell | PowerShell Core
-------------------------------------------------- | ---------- | ----------
Microsoft Windows 10 Home | 5.1 | ?


CHANGELOG.md

# Changelog for Module PSModule.Base"

Started to use as many commonly used defaults to setup a Microsoft PowerShell module where changes and readme are part of.  
Help the PowerShell way is used for sure, but the community also uses some commonly used standards.  
Not sure yet, but maybe these files ( _CHANGELOG.md_ and _README.md_ ) will be small and point to the Microsoft PowerShell way of supplying help.

Links used as reference:

+ [Changelog at Wikipedia](https://en.wikipedia.org/wiki/Markdown)
+ [Changelog CommonMark Spec](http://spec.commonmark.org/)
+ [GitHub markdown based on CommonMark](https://github.github.com/gfm/)
+ [Pandoc - PowerShell module to convert 'Markdown' to ...](http://pandoc.org/)

---

## History

### Unreleased

+ Create and use *module-name.[On, Off]* file for 'write-host' usage because 'write-verbose' doesn't work during 'import-module'.

### 0.4.0 - 210706

+ _Modified_
  + CHANGELOG.md - Modified all inline HTML to VSC MD validation.
  + README.md - Modified all inline HTML to VSC MD validation.
+ _Renamed_
  + Renamed all 'PSModule-Base' to 'PSModule.Base'.

### 0.3.5 - 210706

+ _Added_
  + Invoke-TSPingSweep.ps1

### 0.3.4 - 210209

+ _Modified_
  + Show-ModulePath.ps1
    + Added Parameter 'Version'.
  + Test-WriteProgress.psd1
    + Added Parameter 'Version'.

### 0.3.3 - 210127

Some old modifications on home system that require merge first before i can continue.

+ _Added_
  + PSModule.Base.Off.psd1
+ _Modified_
  + Prefix.ps1
  + PSModule.Base.psd1
+ _Renamed_
  + PSModule.Base-Config-Template.psd1 -> Config-PSModule.Base-Template.psd1
+ _Removed_
  + PSModule.Base.Off

### 0.3.2 - 210127

Pester tests created at the office.

+ *Added*
  + Convert-Bytes.Test.ps1
    + Pester-Convert-Bytes-210127
  + Get-TimeStamp.Test.ps1
    + Pester-Get-TimeStamp-210127

### 0.3.1 - 200807

+ *Added*
  + Test and create folders from configuration.
  + Create variables for the __existing__ folders.

### 0.3.0 - 200414

+ *Modified*
  + Renamed from *PSModule-Personal* to *PSModule.Base*.

### 0.2.9 - 200213

+ *Modified*
  + Moved the module to OneDrive instead of userprofile.

### 0.2.8 - 191129

+ *Modified*
  + Get-Software.ps1
    + Added an object per path for unknown/empty key's.

### 0.2.7 - 191128

+ *Added*
  + Get-Software.ps1
    + Used source: [Gathering Installed Software Using PowerShell](https://mcpmag.com/articles/2017/07/27/gathering-installed-software-using-powershell.aspx) on MCPMAG.
      + Some tweaking to conform to my personal setup.

### 0.2.6 - 191126d

+ *Added*
  + Show-Characterline.ps1
    + Display a line with numbers to show in a terminal as indicator for the width.
    + Default is 80 characters, but use parameter 'MaxLoop' for other numbers.
    + Show-Characterline -MaxLoop 32
      + 012345678901234567890123456789012
    + Returns this string too so if required the total string can be used elsewere.

### 0.2.5 - 191124

+ *Added*
  + Get-Calendar.ps1
    + Something from the internet.
    + A PowerShell way for the *ix 'cal' command.

### 0.2.4 - 191123

+ *Added*
  + Prompt.ps1
    + Copied from my previous personal named module.
    + Changes the prompt. e.g. a red '[Admin]' when the prompt is run as administrator.
  + Test-WriteProgress
    + Copied from my previous personal named module.
    + Multi level demo of progress bars.

### 0.2.3 - 191123

+ *Added*
  + Convert-Bytes : Moved from my previous personal named module.
+ *Modified*
  + Changed version in 'PSModule-Personal.psd1'.

### 0.2.2

+ *Added*
  + Show-Popup.ps1
    + A(nother) wrapper around a Windows popup.
    + This will show a Windows popup with (a) button(s).
    + But in addition it will return a custom object so you can use the return to continue your flow.

### 0.2.1

+ *Added*
  + Get-CommandVersion.ps1
    + The 'Get-Command' showd the version of the module itself. Not of the functions/commands in the module.
    + This function will show the version of the functions within a module.
+ *Modified*
  + Get-TimeStamp

### 0.2.0

+ *Added*
  + PSModule-Personal-Config.psd1
    + A PowerShell Configuration file for this module.
    + In my personal setup I (try to) have a configuration file in each module folder. Where possible I use a configuration file to define values instead of defining values in each script or function. This way function are less prone to errors.
  + Get-TimeStamp.ps1
    + Although in itself this is not to difficult it keeps me from defining a timestamp in every script. This function is in the first module I start so it can be used ASAP.
+ *Modified*
  + PSModule-Personal.psd1
    + ModuleVersion = "0.2.0" : Because of config file.
  + Prefix.ps
    + Temporary modified 'write-verbose' to 'write-host'. The verbose of the module itself doesn't work during import. See [Unrelease](#unrelease).
    + Read the config file and add it to a $Global variable.

### 0.1.1 ( 2019-09-30 )

+ _Added_
  + First setup of this CHANGELOG.md
+ _Modified_
  + README.md

## Examples of types of changes

+ _Added_ : for new features.
+ _Changed_ : for changes in existing functionality.
+ _Deprecated_ : for (soon-to-be) removed features.
+ _Removed_ : for now removed features.
+ _Fixed_ : for any bug fixes.
+ _Security_ : in case of vulnerabilities.


Source/Public/Get-Software.ps1
function Get-Software {
    <#
    .SYNOPSIS
    Get installed software based on the 'uninstall' registry key.

    .DESCRIPTION
    Get installed software based on the 'uninstall' registry key.

    Sources:
    - https://mcpmag.com/articles/2017/07/27/gathering-installed-software-using-powershell.aspx

    Modified the script a bit to comply to the setup for this module.
    Added 'Write-Verbose' since the script didn't run at first. Left it in.
    Added an extra object per path that contains all the empty/unknown key's.

    .EXAMPLE
    PS [Normal] > get-software

    Computername    : <computername>
    Key             : 7-Zip
    DisplayName     : 7-Zip 18.05 (x64)
    Version         : 18.05
    InstallDate     :
    Publisher       : Igor Pavlov
    UninstallString : D:\Software\7-Zip\Uninstall.exe
    InstallLocation : D:\Software\7-Zip\
    InstallSource   :
    HelpLink        :
    EstimatedSizeMB : 4,93

    .EXAMPLE
    PS [Normal] > get-software -Verbose
    VERBOSE: 20191129-145725.792 (F)Get-Software INFO Part: Begin.
    VERBOSE: 20191129-145725.792 (F)Get-Software INFO Part: Process.
    VERBOSE: 20191129-145725.792 (F)Get-Software INFO Check host: MEKD-PC-01
    VERBOSE: 20191129-145725.811 (F)Get-Software INFO Check registry keys.
    VERBOSE: 20191129-145725.820 (F)Get-Software INFO Checking Path: SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall
    VERBOSE: 20191129-145725.831 (F)Get-Software INFO Drilldown into 'Uninstall' key for: HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall
    VERBOSE: 20191129-145725.836 (F)Get-Software INFO Subkeys found # 50
    VERBOSE:  ...Testing Key   1 : 7-Zip
    VERBOSE:  ......DisplayName, Date, Publisher, Displayversion, Uninstallstring, Installlocation, InstallSource, HelpLink, PSCustomObject

    Computername    : <computername>
    Key             : 7-Zip
    DisplayName     : 7-Zip 18.05 (x64)
    Version         : 18.05
    InstallDate     :
    Publisher       : Igor Pavlov
    UninstallString : D:\Software\7-Zip\Uninstall.exe
    InstallLocation : D:\Software\7-Zip\
    InstallSource   :
    HelpLink        :
    EstimatedSizeMB : 4,93

    .EXAMPLE
    PS [Normal] > $AllSw = get-software -Verbose
    VERBOSE: 20191129-145919.810 (F)Get-Software INFO Part: Begin.
    VERBOSE: 20191129-145919.820 (F)Get-Software INFO Part: Process.
    VERBOSE: 20191129-145919.820 (F)Get-Software INFO Check host: <computername>
    VERBOSE: 20191129-145919.880 (F)Get-Software INFO Check registry keys.
    VERBOSE: 20191129-145919.880 (F)Get-Software INFO Checking Path: SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall
    VERBOSE: 20191129-145919.880 (F)Get-Software INFO Drilldown into 'Uninstall' key for: HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall
    VERBOSE: 20191129-145919.880 (F)Get-Software INFO Subkeys found # 50
    VERBOSE:  ...Testing Key   1 : 7-Zip
    VERBOSE:  ......DisplayName, Date, Publisher, Displayversion, Uninstallstring, Installlocation, InstallSource, HelpLink, PSCustomObject    
    ...

    .EXAMPLE
    PS [Normal] > $AllSw = get-software
    PS [Normal] > $AllSw | Where-Object { $_.displayname -eq 'unknown' }

    Computername    : <computername>
    Key             : 20191129-150020.100 (F)Get-Software INFO No information for the following ( 7 ) keys, AddressBook, Connection Manager, 
                        DirectDrawEx, IE40, IE4Data, IE5BAKEX, IEData
    DisplayName     : Unknown
    Version         :
    InstallDate     : 29-11-2019 15:00:20
    Publisher       :
    UninstallString :
    InstallLocation :
    InstallSource   :
    HelpLink        :
    EstimatedSizeMB : 0

    .PARAMETER Version [<SwitchParameter>]
    Will show the version of the script and exit.

    .INPUTS
    The defined parameters.

    .OUTPUTS
    The PSCustomObject will be shown in the default table format.

    .NOTES
    History:
    - 200214 MR
        - V1.1.5 In the 'Begin' part after show version changed 'break' init 'return $scriptversion'.
        - V1.1.6 Removed the 'begin - process - end' parts. Interfered with proper exit '$version'.

    - 191129 MR
        - 1.1.4 Added an object per path that contains all empty/unknown key's.
            Gave them the Displayname 'Unknown'. And the key is ALL the empty/unknown ones.
    
    - 191128 MR
        - 1.0.1 Created this version based on above source.
            Modified parts to comply to my setup.

    #>

    #Region 'Initialize.'   
    # OutputType explained on:
    #   https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_outputtypeattribute?view=powershell-6
    [ OutputType( 'System.Software.Inventory' )]
    
    [ CmdletBinding( )]

    param (
        [ Parameter( )]
        [ Switch ] $Version
        ,
        [ Parameter(
            ValueFromPipeline = $True,
            ValueFromPipelineByPropertyName = $True
        )]
        [ String[ ]] $Computername = $env:COMPUTERNAME
    )
    #EndRegion 'Initialize.'

    #Region 'Block Begin.'
    #Begin {
        # Get name and return version if requested.
        $ScriptName = '(F)' + [io.path]::GetFileNameWithoutExtension( $MyInvocation.MyCommand.Name )
        Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Part: Begin."
        [ version ] $ScriptVersion = '1.1.6'
        if ( $Version ) {
            #Write-Host "$( Get-TimeStamp ) $( $ScriptName ) INFO Test if Version was requested: $( $ScriptVersion )"
            Return $ScriptVersion
        }
    #}
    #EndRegion 'Block Begin.'

    #Region 'Block Process.'
    #Process {
        Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Part: Process."
        [ array ] $NoSoftware = @()

        #Region 'Loop through all computers.'
        ForEach  ( $Computer in  $Computername ) {
            Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Check host: $( $Computer )"
            If ( Test-Connection -ComputerName  $Computer -Count  1 -Quiet ) {
                Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Check registry keys."
                $Paths = @(
                    "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall",
                    "SOFTWARE\\Wow6432node\\Microsoft\\Windows\\CurrentVersion\\Uninstall"
                )

                #Region 'Loop through each path.'
                ForEach ( $Path in $Paths ) {
                    Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Checking Path: $Path"

                    # Create an instance of the Registry Object and open the HKLM base key.
                    Try {
                        $reg = [ microsoft.win32.registrykey ]::OpenRemoteBaseKey(
                            'LocalMachine',
                            $Computer,
                            'Registry64'
                        )
                    }
                    Catch {
                        Write-Error $_
                        Continue
                    }

                    #Region 'Drill down into the Uninstall key using the OpenSubKey Method.'
                    Try {
                        $regkey = $reg.OpenSubKey( $Path )
                        Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Drilldown into 'Uninstall' key for: $( $regkey )"

                        # Retrieve an array of string that contain all the subkey names.
                        $subkeys = $regkey.GetSubKeyNames()
                        $subkeysCount = $subkeys.count
                        Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Subkeys found # $( $subkeysCount )"

                        #Region 'Open each Subkey and use GetValue Method to return the required values for each.'
                        $TempCounter = 1
                        ForEach ( $key in $subkeys ) {
                            $TmpDbg = $DebugSW
                            Write-Verbose "`t...Testing Key $( $TempCounter.ToString().PadLeft( 3 )) : $Key"
                            $thisKey = $Path + "\\" + $key

                            #Region 'Try to get subkey information.'
                            Try {
                                $thisSubKey = $reg.OpenSubKey( $thisKey )
                                
                                # Prevent Objects with empty DisplayName
                                $Tmpvalues = "`t......DisplayName"
                                $DisplayName = $thisSubKey.getValue( "DisplayName" )
                                #If ( $DisplayName -AND $DisplayName -notmatch '^Update  for|rollup|^Security Update|^Service Pack|^HotFix' ) {
                                If ( $DisplayName ) {
                                    $Tmpvalues = $Tmpvalues + ', Date'
                                    $Date = $thisSubKey.GetValue( 'InstallDate' )
                                    If ( $Date ) {
                                        Try {
                                            $Date = [ datetime ]::ParseExact( $Date, 'yyyyMMdd', $Null )
                                        }
                                        Catch {
                                            Write-Warning "$( $Computer ): $_ <$( $Date )>"
                                            $Date = $Null
                                        }
                                    }

                                    # Create New Object with empty Properties.
                                    $Tmpvalues = $Tmpvalues + ', Publisher'
                                    $Publisher = Try {
                                        $thisSubKey.GetValue( 'Publisher' ).Trim()
                                    }
                                    Catch {
                                        $thisSubKey.GetValue( 'Publisher' )
                                    }

                                    $Tmpvalues = $Tmpvalues + ', Displayversion'
                                    $KeyVersion = Try {
                                        # Some weirdness with trailing [char]0 on some strings.
                                        $thisSubKey.GetValue( 'DisplayVersion' ).TrimEnd(([ char[ ]]( 32, 0 )))
                                    }
                                    Catch {
                                        $thisSubKey.GetValue( 'DisplayVersion' )
                                    }

                                    $Tmpvalues = $Tmpvalues + ', Uninstallstring'
                                    $UninstallString = Try {
                                        $thisSubKey.GetValue( 'UninstallString' ).Trim()
                                    }
                                    Catch {
                                        $thisSubKey.GetValue( 'UninstallString' )
                                    }

                                    $Tmpvalues = $Tmpvalues + ', Installlocation'
                                    $InstallLocation = Try {
                                        $thisSubKey.GetValue( 'InstallLocation' ).Trim()
                                    }
                                    Catch {
                                        $thisSubKey.GetValue( 'InstallLocation' )
                                    }

                                    $Tmpvalues = $Tmpvalues + ', InstallSource'
                                    $InstallSource = Try {
                                        $thisSubKey.GetValue( 'InstallSource' ).Trim()
                                    }
                                    Catch {
                                        $thisSubKey.GetValue( 'InstallSource' )
                                    }

                                    $Tmpvalues = $Tmpvalues + ', HelpLink'
                                    $HelpLink = Try {
                                        $thisSubKey.GetValue( 'HelpLink' ).Trim()
                                    }
                                    Catch {
                                        $thisSubKey.GetValue( 'HelpLink' )
                                    }

                                    $Tmpvalues = $Tmpvalues + ', PSCustomObject'
                                    Write-Verbose $Tmpvalues
                                    $Object = [ pscustomobject ] @{
                                        Computername = $Computer
                                        Key = $key
                                        DisplayName = $DisplayName
                                        Version = $KeyVersion
                                        InstallDate = $Date
                                        Publisher = $Publisher
                                        UninstallString = $UninstallString
                                        InstallLocation = $InstallLocation
                                        InstallSource  = $InstallSource
                                        HelpLink = $HelpLink
                                        EstimatedSizeMB = [ decimal ]([ math ]::Round((
                                                $thisSubKey.GetValue( 'EstimatedSize' ) * 1024 ) / 1MB,
                                                2
                                        ))
                                    }
                                    $Object.pstypenames.insert( 0, 'System.Software.Inventory' )
                                    Write-Output $Object
                                }
                                else {
                                    $NoSoftware += $key
                                }
                            }
                            Catch {
                                Write-Warning "$Key : $_"
                            }
                            #EndRegion 'Try to get subkey information.'
                            $TempCounter++
                        }
                        #EndRegion 'Open each Subkey and use GetValue Method to return the required values for each.'

                    }
                    Catch {
                        Write-Warning "$( Get-TimeStamp ) $( $ScriptName ) WARNING Should not come here."
                    }
                    #EndRegion 'Drill down into the Uninstall key using the OpenSubKey Method.'

                    #Region 'Finished all paths, finish.'
                    $reg.Close()
                    $TmpText = "$( Get-TimeStamp ) $( $ScriptName ) INFO No information for the following ( $( $NoSoftware.Count )) keys"
                    foreach ( $NoSW in $NoSoftware ) {
                        $TmpText = $TmpText + ", $( $NoSW )"
                    }
                    Write-Verbose "$( $TmpText )."
                    try {
                        $Object = [ pscustomobject ] @{
                            Computername = $Computer
                            Key = "$( $TmpText )"
                            DisplayName = 'Unknown'
                            Version = ''
                            InstallDate = Get-Date
                            Publisher = ''
                            UninstallString = ''
                            InstallLocation = ''
                            InstallSource  = ''
                            HelpLink = ''
                            EstimatedSizeMB = 0
                        }
                        $Object.pstypenames.insert( 0, 'System.Software.Inventory' )
                        Write-Output $Object                    
                    }
                    catch {
                        Write-Host "$( Get-TimeStamp ) $( $ScriptName ) WARNING Error creating object with empty/unknown key's."
                    }
                    #EndRegion 'Finished all paths, finish.'
                }
                #EndRegion 'Loop through each path.'

            }
            Else {
                Write-Error "$( Get-TimeStamp ) $( $ScriptName ) INFO $($Computer): unable to reach remote system!"
            }
        } 
        #EndRegion 'Loop through all computers.'
    #}
    #EndRegion 'Block Process.'

    #Region 'Block End.'
    #End {
        Write-Verbose "$( Get-TimeStamp ) $( $ScriptName ) INFO Part: End."
    #}
    #EndRegion 'Block End.'
}
